File: dfs-bfs/dfs-cycle/[백준]9466. 텀 프로젝트/solution.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**5)

def dfs(now):
    global no, res
    next = arr[now]
    if v[next] == 0:
        no += 1
        v[next] = no
        dfs(next)
    elif f[next] == 0:
        res += (v[now] - v[next] + 1)
    f[now] = 1
    

t = int(input())
for _ in range(t):
    n = int(input())
    no = 0
    res = 0
    arr = [0] + list(map(int, input().split()))
    v = [0] * (n + 1)
    f = [0] * (n + 1)
    for i in range(1, n + 1):
        if not v[i]:
            no += 1
            v[i] = no
            dfs(i)
    print(n - res)

File: dfs-bfs/dfs-cycle/[백준]9466. 텀 프로젝트/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int t, n, no, res;
int arr[100001], v[100001], f[100001];

void dfs(int n) {
    int next = arr[n];
    if (!v[next]) {
        v[next] = ++no;
        dfs(next);
    }
    else if (!f[next]) {
        res += (v[n] - v[next] + 1);
    }
    f[n] = 1;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(0);
    cin >> t;
    while (t--) {
        memset(v, 0, sizeof(v));
        memset(f, 0, sizeof(f));
        no = res = 0;
        cin >> n;
        for (int i = 1; i <= n; i++) cin >> arr[i];
        for (int i = 1; i <= n; i++) {
            if (!v[i]) {
                v[i] = ++no;
                dfs(i);
            }
        }
        cout << n - res << '\n';
    }
}